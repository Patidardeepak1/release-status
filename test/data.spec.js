const fs = require('fs');
const path = require('path');

const nock = require('nock');

const { getVersionsOrUpdate } = require('../src/data');

function loadFixture(filePath) {
  return JSON.parse(fs.readFileSync(path.join(__dirname, 'fixtures', filePath), 'utf-8'));
}

const chromiumM122Fixture = loadFixture('chromium-m122.json');
const chromiumM124Fixture = loadFixture('chromium-m124.json');
const chromiumM126Fixture = loadFixture('chromium-m126.json');
const chromiumM128Fixture = loadFixture('chromium-m128.json');
const chromiumM130Fixture = loadFixture('chromium-m130.json');
const indexJsonFixture = loadFixture('index.json');

nock('https://electronjs.org').get('/headers/index.json').reply(200, indexJsonFixture);

nock('https://chromiumdash.appspot.com')
  .get('/fetch_milestone_schedule')
  .query({ mstone: '122' })
  .reply(200, chromiumM122Fixture)
  .get('/fetch_milestone_schedule')
  .query({ mstone: '124' })
  .reply(200, chromiumM124Fixture)
  .get('/fetch_milestone_schedule')
  .query({ mstone: '126' })
  .reply(200, chromiumM126Fixture)
  .get('/fetch_milestone_schedule')
  .query({ mstone: '128' })
  .reply(200, chromiumM128Fixture)
  .get('/fetch_milestone_schedule')
  .query({ mstone: '130' })
  .reply(200, chromiumM130Fixture);

describe('getVersionsOrUpdate', () => {
  beforeAll(() => {
    nock.disableNetConnect();
  });

  afterAll(() => {
    nock.enableNetConnect();
  });

  describe('majors', () => {
    const isStable = ({ isStable }) => isStable === true;
    const isSupported = ({ isSupported }) => isSupported === true;
    const isNotSupported = ({ isSupported }) => isSupported === false;

    it('calculated dates match snapshot', async () => {
      const { majors } = await getVersionsOrUpdate();
      expect(
        majors.map(({ major, alpha, beta, stable, endOfLife }) => ({
          major,
          alpha,
          beta,
          stable,
          endOfLife,
        })),
      ).toMatchSnapshot();
    });

    it('has all majors from E3 onward', async () => {
      const { majors } = await getVersionsOrUpdate();

      expect(majors.map(({ major }) => major)).toEqual([...Array(31).keys()].slice(2).reverse());
    });

    it('has alphas only from E15 onward', async () => {
      const { majors } = await getVersionsOrUpdate();
      const alphaIsNull = ({ alpha }) => alpha === null;

      expect(majors.filter(({ major }) => major < 15).every(alphaIsNull)).toEqual(true);
      expect(majors.filter(({ major }) => major >= 15).some(alphaIsNull)).toEqual(false);
    });

    it('all majors before E29 are stable', async () => {
      const { majors } = await getVersionsOrUpdate();

      expect(majors.filter(({ major }) => major < 29).every(isStable)).toEqual(true);
    });

    it('all majors from E29 onward are not stable', async () => {
      const { majors } = await getVersionsOrUpdate();

      expect(majors.filter(({ major }) => major >= 29).some(isStable)).toEqual(false);
    });

    it('all majors before E26 are not supported', async () => {
      const { majors } = await getVersionsOrUpdate();

      expect(majors.filter(({ major }) => major < 26).some(isSupported)).toEqual(false);
    });

    it('all majors from E26 onward are supported', async () => {
      const { majors } = await getVersionsOrUpdate();

      expect(majors.filter(({ major }) => major >= 26).every(isSupported)).toEqual(true);
    });

    it('all stable majors from E15 onward have an actual date for alpha', async () => {
      const { majors } = await getVersionsOrUpdate();

      expect(
        majors
          .filter(({ major }) => major >= 15)
          .filter(isStable)
          .every(({ alpha }) => {
            return typeof alpha.actual === 'string';
          }),
      ).toEqual(true);
    });

    it('all stable majors have actual dates for beta/stable', async () => {
      const { majors } = await getVersionsOrUpdate();

      expect(
        majors.filter(isStable).every(({ beta, stable }) => {
          return typeof beta.actual === 'string' && typeof stable.actual === 'string';
        }),
      ).toEqual(true);
    });

    it('all stable majors from E15 onward have even Chromium milestones', async () => {
      const { majors } = await getVersionsOrUpdate();

      expect(
        majors
          .filter(({ major }) => major >= 15)
          .filter(isStable)
          .every(({ chromium }) => chromium % 2 === 0),
      ).toEqual(true);
    });

    it('all unsupported majors have actual dates for endOfLife', async () => {
      const { majors } = await getVersionsOrUpdate();

      expect(
        majors.filter(isNotSupported).every(({ endOfLife }) => {
          return typeof endOfLife.actual === 'string';
        }),
      ).toEqual(true);
    });

    it('all supported majors have estimated dates for endOfLife', async () => {
      const { majors } = await getVersionsOrUpdate();

      expect(
        majors.filter(isSupported).every(({ endOfLife }) => {
          return typeof endOfLife.estimated === 'string';
        }),
      ).toEqual(true);
    });

    it.todo('falls back to stored data on Chromium Dash error');
  });
});
