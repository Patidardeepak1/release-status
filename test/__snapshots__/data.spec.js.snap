// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getVersionsOrUpdate majors calculated dates match snapshot 1`] = `
[
  {
    "alpha": {
      "estimated": "2024-02-21",
    },
    "beta": {
      "estimated": "2024-03-20",
    },
    "endOfLife": {
      "estimated": "2024-10-15",
    },
    "major": 30,
    "stable": {
      "estimated": "2024-04-16",
    },
  },
  {
    "alpha": {
      "actual": "2023-12-06",
    },
    "beta": {
      "estimated": "2024-01-24",
    },
    "endOfLife": {
      "estimated": "2024-08-20",
    },
    "major": 29,
    "stable": {
      "estimated": "2024-02-20",
    },
  },
  {
    "alpha": {
      "actual": "2023-10-11",
    },
    "beta": {
      "actual": "2023-10-30",
    },
    "endOfLife": {
      "estimated": "2024-06-11",
    },
    "major": 28,
    "stable": {
      "actual": "2023-12-04",
    },
  },
  {
    "alpha": {
      "actual": "2023-08-16",
    },
    "beta": {
      "actual": "2023-09-11",
    },
    "endOfLife": {
      "estimated": "2024-04-16",
    },
    "major": 27,
    "stable": {
      "actual": "2023-10-09",
    },
  },
  {
    "alpha": {
      "actual": "2023-05-31",
    },
    "beta": {
      "actual": "2023-06-26",
    },
    "endOfLife": {
      "estimated": "2024-02-20",
    },
    "major": 26,
    "stable": {
      "actual": "2023-08-14",
    },
  },
  {
    "alpha": {
      "actual": "2023-04-06",
    },
    "beta": {
      "actual": "2023-05-01",
    },
    "endOfLife": {
      "actual": "2023-12-04",
    },
    "major": 25,
    "stable": {
      "actual": "2023-05-29",
    },
  },
  {
    "alpha": {
      "actual": "2023-02-09",
    },
    "beta": {
      "actual": "2023-03-06",
    },
    "endOfLife": {
      "actual": "2023-10-09",
    },
    "major": 24,
    "stable": {
      "actual": "2023-04-03",
    },
  },
  {
    "alpha": {
      "actual": "2022-12-05",
    },
    "beta": {
      "actual": "2023-01-10",
    },
    "endOfLife": {
      "actual": "2023-08-14",
    },
    "major": 23,
    "stable": {
      "actual": "2023-02-06",
    },
  },
  {
    "alpha": {
      "actual": "2022-09-28",
    },
    "beta": {
      "actual": "2022-10-25",
    },
    "endOfLife": {
      "actual": "2023-10-10",
    },
    "major": 22,
    "stable": {
      "actual": "2022-11-29",
    },
  },
  {
    "alpha": {
      "actual": "2022-08-03",
    },
    "beta": {
      "actual": "2022-08-30",
    },
    "endOfLife": {
      "actual": "2023-04-03",
    },
    "major": 21,
    "stable": {
      "actual": "2022-09-26",
    },
  },
  {
    "alpha": {
      "actual": "2022-05-25",
    },
    "beta": {
      "actual": "2022-06-21",
    },
    "endOfLife": {
      "actual": "2023-02-06",
    },
    "major": 20,
    "stable": {
      "actual": "2022-08-01",
    },
  },
  {
    "alpha": {
      "actual": "2022-03-30",
    },
    "beta": {
      "actual": "2022-04-28",
    },
    "endOfLife": {
      "actual": "2022-11-29",
    },
    "major": 19,
    "stable": {
      "actual": "2022-05-23",
    },
  },
  {
    "alpha": {
      "actual": "2022-02-02",
    },
    "beta": {
      "actual": "2022-03-08",
    },
    "endOfLife": {
      "actual": "2022-09-26",
    },
    "major": 18,
    "stable": {
      "actual": "2022-03-29",
    },
  },
  {
    "alpha": {
      "actual": "2021-11-17",
    },
    "beta": {
      "actual": "2022-01-06",
    },
    "endOfLife": {
      "actual": "2022-08-01",
    },
    "major": 17,
    "stable": {
      "actual": "2022-02-01",
    },
  },
  {
    "alpha": {
      "actual": "2021-09-22",
    },
    "beta": {
      "actual": "2021-10-20",
    },
    "endOfLife": {
      "actual": "2022-05-23",
    },
    "major": 16,
    "stable": {
      "actual": "2021-11-16",
    },
  },
  {
    "alpha": {
      "actual": "2021-07-21",
    },
    "beta": {
      "actual": "2021-08-31",
    },
    "endOfLife": {
      "actual": "2022-05-23",
    },
    "major": 15,
    "stable": {
      "actual": "2021-09-21",
    },
  },
  {
    "alpha": null,
    "beta": {
      "actual": "2021-05-26",
    },
    "endOfLife": {
      "actual": "2022-03-29",
    },
    "major": 14,
    "stable": {
      "actual": "2021-08-30",
    },
  },
  {
    "alpha": null,
    "beta": {
      "actual": "2021-03-05",
    },
    "endOfLife": {
      "actual": "2022-02-01",
    },
    "major": 13,
    "stable": {
      "actual": "2021-05-25",
    },
  },
  {
    "alpha": null,
    "beta": {
      "actual": "2020-11-19",
    },
    "endOfLife": {
      "actual": "2021-11-16",
    },
    "major": 12,
    "stable": {
      "actual": "2021-03-02",
    },
  },
  {
    "alpha": null,
    "beta": {
      "actual": "2020-08-26",
    },
    "endOfLife": {
      "actual": "2021-08-30",
    },
    "major": 11,
    "stable": {
      "actual": "2020-11-16",
    },
  },
  {
    "alpha": null,
    "beta": {
      "actual": "2020-05-22",
    },
    "endOfLife": {
      "actual": "2021-05-25",
    },
    "major": 10,
    "stable": {
      "actual": "2020-08-24",
    },
  },
  {
    "alpha": null,
    "beta": {
      "actual": "2020-02-10",
    },
    "endOfLife": {
      "actual": "2021-03-02",
    },
    "major": 9,
    "stable": {
      "actual": "2020-05-19",
    },
  },
  {
    "alpha": null,
    "beta": {
      "actual": "2019-10-24",
    },
    "endOfLife": {
      "actual": "2020-11-16",
    },
    "major": 8,
    "stable": {
      "actual": "2020-02-03",
    },
  },
  {
    "alpha": null,
    "beta": {
      "actual": "2019-08-01",
    },
    "endOfLife": {
      "actual": "2020-08-24",
    },
    "major": 7,
    "stable": {
      "actual": "2019-10-21",
    },
  },
  {
    "alpha": null,
    "beta": {
      "actual": "2019-05-01",
    },
    "endOfLife": {
      "actual": "2020-05-19",
    },
    "major": 6,
    "stable": {
      "actual": "2019-07-29",
    },
  },
  {
    "alpha": null,
    "beta": {
      "actual": "2019-01-24",
    },
    "endOfLife": {
      "actual": "2020-02-03",
    },
    "major": 5,
    "stable": {
      "actual": "2019-04-24",
    },
  },
  {
    "alpha": null,
    "beta": {
      "actual": "2018-10-11",
    },
    "endOfLife": {
      "actual": "2019-10-21",
    },
    "major": 4,
    "stable": {
      "actual": "2018-12-20",
    },
  },
  {
    "alpha": null,
    "beta": {
      "actual": "2018-06-20",
    },
    "endOfLife": {
      "actual": "2019-07-29",
    },
    "major": 3,
    "stable": {
      "actual": "2018-09-18",
    },
  },
  {
    "alpha": null,
    "beta": {
      "actual": "2018-02-21",
    },
    "endOfLife": {
      "actual": "2019-04-24",
    },
    "major": 2,
    "stable": {
      "actual": "2018-05-01",
    },
  },
]
`;
